wallet/datastore.go

```diff
@@ -0,0 +1,318 @@
+	return &wallet, err
+}
+
+// GetByProviderLinkingID gets a wallet by a provider address
+func (pg *Postgres) GetByProviderLinkingID(providerLinkingID uuid.UUID) (*[]walletutils.Info, error) {
+	statement := `
+	select
+		id, provider, provider_id, public_key, provider_linking_id, anonymous_address,
+		user_deposit_account_provider, user_deposit_account_provider_id,
+		user_deposit_account_anonymous_address
+	from
+		wallets
+	WHERE provider_linking_id = $1
+	`
+	var wallets []walletutils.Info
+	err := pg.RawDB().Select(&wallets, statement, providerLinkingID)
+	return &wallets, err
+}
+
+// InsertWallet inserts the given wallet
+func (pg *Postgres) InsertWallet(wallet *walletutils.Info) error {
+	// NOTE on conflict do nothing because none of the wallet information is updateable
+	statement := `
+	INSERT INTO wallets (id, provider, provider_id, public_key)
+	VALUES ($1, $2, $3, $4)
+	ON CONFLICT DO NOTHING`
+	_, err := pg.RawDB().Exec(
+		statement,
+		wallet.ID,
+		wallet.Provider,
+		wallet.ProviderID,
+		wallet.PublicKey,
+	)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// SetAnonymousAddress sets the anon addresses of the provided wallets
+func (pg *Postgres) SetAnonymousAddress(ID string, anonymousAddress *uuid.UUID) error {
+	statement := `
+	UPDATE wallets
+	SET
+			anonymous_address = $2
+	WHERE id = $1
+	`
+	_, err := pg.RawDB().Exec(
+		statement,
+		ID,
+		anonymousAddress,
+	)
+	return err
+}
+
+// TxLinkWalletInfo pass a tx to set the anonymous address
+func (pg *Postgres) TxLinkWalletInfo(tx *sqlx.Tx, ID string, providerLinkingID uuid.UUID, anonymousAddress *uuid.UUID) error {
+	statement := `
+	UPDATE wallets
+	SET
+			provider_linking_id = $2,
+			anonymous_address = $3
+	WHERE id = $1;`
+	r, err := tx.Exec(
+		statement,
+		ID,
+		providerLinkingID,
+		anonymousAddress,
+	)
+	count, _ := r.RowsAffected()
+	if count < 1 {
+		return errors.New("should have updated at least one wallet")
+	}
+	return err
+}
+
+func txGetByProviderLinkingID(tx *sqlx.Tx, providerLinkingID uuid.UUID) (*[]walletutils.Info, error) {
+	statement := `
+	select
+		id, provider, provider_id, public_key, provider_linking_id, anonymous_address,
+		user_deposit_account_provider, user_deposit_account_provider_id,
+		user_deposit_account_anonymous_address
+	from
+		wallets
+	WHERE provider_linking_id = $1
+	`
+	var wallets []walletutils.Info
+	err := tx.Select(&wallets, statement, providerLinkingID)
+	return &wallets, err
+}
+
+// LinkWallet links a wallet together
+func (pg *Postgres) LinkWallet(ID string, providerLinkingID uuid.UUID, anonymousAddress *uuid.UUID) error {
+	tx, err := pg.RawDB().Beginx()
+	if err != nil {
+		return err
+	}
+	defer pg.RollbackTx(tx)
+
+	walletsMatchingProviderLinkingID, err := txGetByProviderLinkingID(tx, providerLinkingID)
+	if err != nil {
+		return errorutils.Wrap(err, "error looking up wallets by provider id")
+	}
+	walletLinkedLength := len(*walletsMatchingProviderLinkingID)
+	if walletLinkedLength >= 3 {
```

Comment:
we should change this to 4 to match current ledger code

Reply:
makes sense

Reply:
:+1:

Reply:
https://github.com/brave-intl/bat-go/pull/314/commits/ee91546b83628e93123027d39ade0ef864f1a4bb

Reply:
actually done here: https://github.com/brave-intl/bat-go/pull/314/commits/cc63d3438d046dbf87e065e0a17c4971c5c831f4